### ISSUE TITLE
Code Organization

### ISSUE THREAD

**Issue by doakey3** [GDquest/Blender-power-sequencer#101]
I propose changing the code organization of this addon.

**Guidelines**

* Where possible, separate large parts of code into libraries
* Each operator (and any functions used only by itself) in a separate
  python module. Try to avoid modules with 1000+ lines.
* Each shared function in it's own module inside a "utils" library, accessible by all operators that use it.
* Each operator named with camel case naming convention.
* Each python module named with all lowercase letters with "_" separating words.

See [Blender's Addon Guidelines](https://wiki.blender.org/index.php/Dev:Py/Scripts/Guidelines/Addons)

Here's an example of how the addon might be organized.

```
Blender-power-sequencer/ 
├── __init__.py 
├── .gitignore
├── README.md 
├── operators/ 
|   ├── __init__.py 
|   ├── ripple_delete.py
|   └── utils/ 
|       └──a_shared_function.py 
├── updater/ 
|   ├── __init__.py 
|   ├── check_update.py 
|   └── utils/ 
|       └──singleton_updater.py
```

**Comment by NathanLovato** [comment]
Hey, thanks Daniel. I appreciate proposals, but I also need context for them. Do you mind sharing the reason why you organize everything into packages? I guess to see all the operators in your file browser. Note I don't mind, I just want to understand the benefit(s) as it's extra work and I have more on my plate than I can handle.

Be it the official add-ons or the Blender sources themselves, each module can be quite long and that's neither hard to search nor to maintain. This add-on already uses yapf for formatting, is split into modules and follows the official guidelines, maybe one or two old blocks of code aside.

The add-on updater is an external lib we'll likely have to update from upstream so we shouldn't touch it. But then it can go in an external/ package or something where we put libs that'd break conventions.

There's remnants of the initial dirty set of script this tool was lying around, so there's definitely some cleanup to do. Lately my priority has been to refine the UX.

In any case I don't mind too much, and I'm okay if it can help other people better understand the code. But first I'd rather remove as many shared functions as possible, move them inside of operator classes, even at the cost of a little code duplication (when otherwise the util function needs too many arguments to support the specific needs of 2-3 different classes). Did a little bit of this already but there's some work left to do.

@davcri Do you mind sharing your thoughts? As someone who may not know the codebase by heart does it feel clearer with the folder structure above?

**Comment by NathanLovato** [comment]
Note it will answer #6

**Comment by davcri** [comment]
> In any case I don't mind too much, and I'm okay if it can help other people better understand the code. 

Splitting operators in several files seems more elegant to me (from a generic object oriented design perspective). Doing it right now can be a bit daunting, but I think that this will be a good move for the future, especially if the add-on will grow bigger and bigger.

In doubt, we can create a *refactoring* branch and then evaluate pros and cons.

> But first I'd rather remove as many shared functions as possible, move them inside of operator classes, even at the cost of a little code duplication (when otherwise the util function needs too many arguments to support the specific needs of 2-3 different classes). Did a little bit of this already but there's some work left to do.

I understand that this is a quick solution for the current state of the add-on, but usually this is discouraged: if the add-on will grow (and I hope it will), there will be a lot of duplicate code and then it will be a problem.  

I don't know which specific functions/operators we are talking about, but we can study a more flexible solution and see if it does fit better.

**Comment by NathanLovato** [comment]
Alright, waiting for Daniel's reply and why not then :)

> I don't know which specific functions/operators we are talking about

Not every function, it's just I wrote some initially that made it more complicated to for different objects to use them. I think there's still a few left. Learned this from game development: in large code-base you tend to see this "duplication", because you're often better off having a class with its own, very specific method than adding more parameters and complexity to a shared function.

The idea is just to go through the code again and remove the last few "fake utils" before changing the project structure.

**Comment by doakey3** [comment]
Organization and code styling is a sticky subject and my opinions are not perfect, so I'm totally fine if this issue's ignored completely. Also, I made this issue after only a few minutes of looking at the code. 

I forked this code thinking I might add something, but I found it difficult to figure out which modules held which operators. For instance, if I write a new operator, should I put it into "simple_operations.py", or "video_editing.py". The distinction isn't obvious.

If operators are separated into different modules, I can easily find the operator I want to edit without running `grep -lr "operator_name"` to find where it is.

Importing is ambiguous at first glance. For example does "export.py" rely on functions from "proxy.py?" Well, no. But I had to open export.py first to see. If code is organized into libraries and sub libraries, I can infer, based on code organization alone, which modules may be importing from others. (Typically, library modules won't import from modules in parent directories).

Bottom line is I got a little lost when reading through the code, but I can't rule out if that's just because I'm dumb.

P.S. You're right about the addon_updater modules. I didn't know about addon_updating until I found it here. I'm going to implement this in my own addons.

**Comment by NathanLovato** [comment]
> so I'm totally fine if this issue's ignored completely

Nah not at all, I was hoping for someone to suggest some better code organization :)
The idea atm is that each module is self-contained. However their names are a bit random ;). I get the point, and okay, let's do this - let's even do this for 1.0. I'll do a bit of communication with the release to hopefully motivate some devs to give a hand - an accessible and clean structure will definitely help people get started.

I'll just add a task to check and remove superfluous shared functions and we can do this.

**Comment by doakey3** [comment]
Each module has a descriptive name about what it does. Supportive functions are stowed away in a library. You use docstrings to describe functions and it looks like you're using a consistent style in the code. Looks like the project will be easier to understand and maintain. I'll have to take a look at the code this weekend and see if there's anything I can contribute to this project.

**Comment by davcri** [comment]
> Took care of the first step this evening - is this what we're aiming for? https://github.com/GDquest/Blender-power-sequencer/tree/master/operators

Great, I like the new structure! Maybe we could share some part of the docstrings with the actual documentation ?

**Comment by NathanLovato** [comment]
Thanks! Gonna do it one step at a time when taking a break from work.

@davcri With this structure we can do it, but it'd be best to have all the docs in one place. Two locations is enough to scare contributors away from Godot's docs and to confuse a lot of people. So either have everything in the source code or nothing at all. 1-line descriptions wouldn't help too much imo, if it wouldn't waste time.

Generating and updating part of the docs structure from the operators folder is possible though. Just so that we have a comprehensive list of features we know we have to write docs for. Now one issue left is for the system to be really useful and scalable, if we parse the sources' structure we should use it to organize the docs as well. Does it seem okay if we do this?

```
├── operators/ 
|   └── mouse_editing/
|       ├── mouse_cut.py
|       └── mouse_trim.py
|   └── keyboard_editing/
|       ├── ripple_delete.py
|       └── ...
```

There are over 40 operators already in the add-on and more to come so it may be worth organizing operators in sub-packages? This way 1 package == 1 page and 1 TOC in the docs.

**Comment by NathanLovato** [comment]
For the handlers, are you okay with a /handlers folder, separate from operators? There's no dependencies on other packages, at least for now, but there may be imports from ..operators.utils if we were to add auto-updating fades for instance

**Comment by doakey3** [comment]
I think breaking up large pieces of code into more digestible pieces is generally a good idea. I get a little queasy  when libraries import from parent directories, but this is probably one of those cases where [practicality beats purity](https://www.python.org/dev/peps/pep-0020/). I noticed the [retopoflow addon](https://github.com/CGCookie/retopoflow) has a top-level lib and lower level modules import from it.

**Comment by NathanLovato** [comment]
Calling it a day with https://github.com/GDquest/Blender-power-sequencer/commit/50b53e68c82f6112e94db7134c8321775856c5db
